File: src/ros2_bridge/mappers/foxglove_msgs/circle_annotation.rs
//! Typed mapper for `foxglove_msgs/msg/CircleAnnotation`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn circle_annotation_to_bus(msg: ros_env::foxglove_msgs::msg::CircleAnnotation) -> crate::foxglove_msgs::msg::v1::CircleAnnotation {
    crate::foxglove_msgs::msg::v1::CircleAnnotation {
        timestamp: Some(crate::ros2_bridge::mappers::convert::time_to_timestamp(msg.timestamp)),
        position: Some(crate::ros2_bridge::mappers::foxglove_msgs::point2::point2_to_bus(msg.position)),
        diameter: msg.diameter,
        thickness: msg.thickness,
        fill_color: Some(crate::ros2_bridge::mappers::foxglove_msgs::color::color_to_bus(msg.fill_color)),
        outline_color: Some(crate::ros2_bridge::mappers::foxglove_msgs::color::color_to_bus(msg.outline_color)),
        metadata: msg.metadata.into_iter().map(crate::ros2_bridge::mappers::foxglove_msgs::key_value_pair::key_value_pair_to_bus).collect(),
    }
}

pub(crate) fn circle_annotation_to_ros(bus: crate::foxglove_msgs::msg::v1::CircleAnnotation) -> ros_env::foxglove_msgs::msg::CircleAnnotation {
    ros_env::foxglove_msgs::msg::CircleAnnotation {
        timestamp: crate::ros2_bridge::mappers::convert::timestamp_to_time(bus.timestamp.unwrap_or_default()),
        position: crate::ros2_bridge::mappers::foxglove_msgs::point2::point2_to_ros(bus.position.unwrap_or_default()),
        diameter: bus.diameter,
        thickness: bus.thickness,
        fill_color: crate::ros2_bridge::mappers::foxglove_msgs::color::color_to_ros(bus.fill_color.unwrap_or_default()),
        outline_color: crate::ros2_bridge::mappers::foxglove_msgs::color::color_to_ros(bus.outline_color.unwrap_or_default()),
        metadata: bus.metadata.into_iter().map(crate::ros2_bridge::mappers::foxglove_msgs::key_value_pair::key_value_pair_to_ros).collect(),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct FoxgloveMsgsCircleAnnotationMapper;

impl TypedTopicMapper for FoxgloveMsgsCircleAnnotationMapper {
    type Ros = ros_env::foxglove_msgs::msg::CircleAnnotation;
    type Bus = crate::foxglove_msgs::msg::v1::CircleAnnotation;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(circle_annotation_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(circle_annotation_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/foxglove_msgs/circle_annotation.py
"""Generated mapper for `foxglove_msgs/msg/CircleAnnotation`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.foxglove_msgs.point2 import point2_to_bus, point2_to_ros
from robot_bus.ros2_bridge.mappers.foxglove_msgs.color import color_to_bus, color_to_ros
from robot_bus.ros2_bridge.mappers.foxglove_msgs.key_value_pair import key_value_pair_to_bus, key_value_pair_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.foxglove_msgs.msg.v1 import CircleAnnotation as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def circle_annotation_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.timestamp = _convert.time_to_timestamp(msg.timestamp)
    bus.position.CopyFrom(point2_to_bus(msg.position))
    bus.diameter = msg.diameter
    bus.thickness = msg.thickness
    bus.fill_color.CopyFrom(color_to_bus(msg.fill_color))
    bus.outline_color.CopyFrom(color_to_bus(msg.outline_color))
    bus.metadata.extend([key_value_pair_to_bus(x) for x in msg.metadata])
    return bus


def circle_annotation_to_ros(bus):
    from foxglove_msgs.msg import CircleAnnotation as RosMsg

    out = RosMsg()
    out.timestamp = _convert.timestamp_to_time(bus.timestamp)
    out.position = point2_to_ros(bus.position)
    out.diameter = bus.diameter
    out.thickness = bus.thickness
    out.fill_color = color_to_ros(bus.fill_color)
    out.outline_color = color_to_ros(bus.outline_color)
    out.metadata = [key_value_pair_to_ros(x) for x in bus.metadata]
    return out


class FoxgloveMsgsCircleAnnotationMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from foxglove_msgs.msg import CircleAnnotation as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return circle_annotation_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return circle_annotation_to_ros(bus)
